apiserver/facades/client/application/application_unit_test.go

```diff
@@ -1126,6 +1131,272 @@ func (s *ApplicationSuite) TestDeployCharmOrigin(c *gc.C) {
 	c.Assert(s.deployParams["hub"].CharmOrigin.Source, gc.Equals, corecharm.Source("charm-hub"))
 }
 
+func createCharmOriginFromURL(c *gc.C, curl *charm.URL) *params.CharmOrigin {
+	switch curl.Schema {
+	case "cs":
+		return &params.CharmOrigin{Source: "charm-store"}
+	case "local":
+		return &params.CharmOrigin{Source: "local"}
+	default:
+		return &params.CharmOrigin{Source: "charm-hub"}
+	}
+}
+
+func (s *ApplicationSuite) TestApplicationDeployWithStorage(c *gc.C) {
+	ctrl := s.setup(c)
+	defer ctrl.Finish()
+
+	ch := s.expectCharm(ctrl, &charm.Meta{Storage: map[string]charm.Storage{
+		"data":    {Type: charm.StorageBlock},
+		"allecto": {Type: charm.StorageBlock},
+	}}, nil, &charm.Config{})
+	curl := charm.MustParseURL("cs:utopic/storage-block-10")
+	s.backend.EXPECT().Charm(curl).Return(ch, nil)
+
+	storageConstraints := map[string]storage.Constraints{
+		"data": {
+			Count: 1,
+			Size:  1024,
+			Pool:  "modelscoped-block",
+		},
+	}
+	args := params.ApplicationDeploy{
+		ApplicationName: "my-app",
+		CharmURL:        curl.String(),
+		CharmOrigin:     createCharmOriginFromURL(c, curl),
+		NumUnits:        1,
+		Storage:         storageConstraints,
+	}
+	results, err := s.api.Deploy(params.ApplicationsDeploy{
+		Applications: []params.ApplicationDeploy{args}},
+	)
+
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results, gc.HasLen, 1)
+	c.Assert(results.Results[0].Error, gc.IsNil)
+
+	c.Assert(s.deployParams["my-app"].Storage, gc.DeepEquals, storageConstraints)
+}
+
+func (s *ApplicationSuite) TestApplicationDeployDefaultFilesystemStorage(c *gc.C) {
+	ctrl := s.setup(c)
+	defer ctrl.Finish()
+
+	ch := s.expectCharm(ctrl, &charm.Meta{Storage: map[string]charm.Storage{
+		"data": {Type: charm.StorageFilesystem, ReadOnly: true},
+	}}, nil, &charm.Config{})
+	curl := charm.MustParseURL("cs:utopic/storage-filesystem-1")
+	s.backend.EXPECT().Charm(curl).Return(ch, nil)
+
+	args := params.ApplicationDeploy{
+		ApplicationName: "my-app",
+		CharmURL:        curl.String(),
+		CharmOrigin:     createCharmOriginFromURL(c, curl),
+		NumUnits:        1,
+	}
+	results, err := s.api.Deploy(params.ApplicationsDeploy{
+		Applications: []params.ApplicationDeploy{args}},
+	)
+
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results, gc.HasLen, 1)
+	c.Assert(results.Results[0].Error, gc.IsNil)
+}
+
+func (s *ApplicationSuite) expectMachine(ctrl *gomock.Controller) *mocks.MockMachine {
```

Comment:
No actual machine.Expect() is called in this method.  Is it really required?

Reply:
I wrote this expecting that I would use it to populate some EXPECTS in the future, but that never happened. Given that, and how small the Machine interface is, I think you're right this can be dropped